src/casanovoutils/downsample.py

```
@@ -63,6 +66,90 @@ def downsample_spectra(
+    Sample up to k spectra per peptide using reservoir sampling.
+
+    Makes a single streaming pass through *spectra*, maintaining a reservoir
+    of size k per unique peptide sequence.  For the j-th occurrence of a
+    peptide:
+
+    * if j <= k: add the spectrum to the reservoir unconditionally.
+    * if j > k: replace a uniformly random reservoir slot with probability
+      k/j (i.e. draw a random integer in ``[0, j-1]``; if it falls within
+      ``[0, k-1]``, replace that slot).
+
+    Memory usage is O(unique peptides × k) rather than O(total spectra).
```

Ruff (0.15.9)

[warning] Docstring contains ambiguous `×` (MULTIPLICATION SIGN). Did you mean `x` (LATIN SMALL LETTER X)?

(RUF002)